proof/crefine/RISCV64/Schedule_C.thy

```diff
@@ -593,35 +593,957 @@ lemma length_scRefills_bounded:
+       apply (clarsimp simp: sc_ptr_to_crefill_ptr_def typ_heap_simps csched_context_relation_def)
+      apply (clarsimp simp: crefill_relation_def csched_context_relation_def)
+      apply (frule rf_sr_refill_buffer_relation)
+      apply (frule h_t_valid_clift_Some_iff[THEN iffD1])
+      apply (clarsimp cong: StateSpace.state.fold_congs)
+      apply (rule_tac old_sc=sc and n="scRefillTail sc"
+                      and fa="rAmount_update (\<lambda>_. rAmount rv + rAmount new)"
+                       in rf_sr_refill_update2; fastforce?)
+         apply (fastforce simp: sched_context.expand)
+        apply (clarsimp simp: typ_heap_simps)
+        apply (fastforce simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def crefill_relation_def)
+       apply (fastforce simp: csched_context_relation_def)
+      apply (fastforce dest: crefill_relationD
+                       simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def crefill_relation_def)
+     apply (ctac add: refill_full_ccorres)
+       apply (rule ccorres_cond[where R=\<top>])
+         apply(clarsimp simp: to_bool_def)
+        apply (ctac add: refill_add_tail_ccorres)
+       apply (rule ccorres_rhs_assoc)
+       apply (rule ccorres_split_nothrow)
+           apply (clarsimp simp: setRefillTl_def updateRefillTl_def)
+           apply (rule updateSchedContext_ccorres_lemma3
+                        [where P="\<lambda>sc. scRefillTail sc < length (scRefills sc)" and P'="?abs"])
+             apply vcg
+            apply fastforce
+           apply clarsimp
+           apply (rename_tac sc sc')
+           apply (frule (1) sc_ko_at_valid_objs_valid_sc'[OF _ invs_valid_objs'])
+           apply (frule rf_sr_refill_buffer_relation)
+           apply (frule_tac n="scRefillTail sc" in h_t_valid_refill, fastforce+)
+           apply (intro conjI)
+              apply (clarsimp simp: typ_heap_simps)
+             apply (rule disjCI2)
+             apply (rule_tac n="length (scRefills sc)" in array_assertion_shrink_right)
+              apply (fastforce intro: sc_at_array_assertion')
+             apply (clarsimp simp: typ_heap_simps csched_context_relation_def)
+            apply (clarsimp simp: sc_ptr_to_crefill_ptr_def typ_heap_simps csched_context_relation_def)
+           apply (clarsimp simp: crefill_relation_def csched_context_relation_def)
+           apply (frule rf_sr_refill_buffer_relation)
+           apply (frule h_t_valid_clift_Some_iff[THEN iffD1])
+           apply (clarsimp cong: StateSpace.state.fold_congs)
+           apply (rule_tac old_sc=sc and n="scRefillTail sc"
+                           and fa="rTime_update (\<lambda>_. rTime new - rAmount rv)"
+                            in rf_sr_refill_update2; fastforce?)
+              apply (fastforce simp: sched_context.expand)
+             apply (clarsimp simp: typ_heap_simps)
+             apply (fastforce simp: typ_heap_simps' sc_ptr_to_crefill_ptr_def crefill_relation_def)
```

Comment:
unusual having typ_heap_simps in two commands like this; is this proof copied?

Reply:
The `clarsimp` uses `typ_heap_simps` and then the `fastforce` uses `typ_heap_simps'`. I'm quite sure that neither can be removed, and I think I have also tried without success before to condense it into one command. It would be cool to see whether this could be improved, or to properly understand what's going on, but I don't think I could figure it out before

Reply:
And yes the proof was copied, and it appears like this in a few places, now

Reply:
We can leave it then. If it were new, would be worth more investigation.